### DISCUSSION TITLE
Support ProMotion on iPhone 13 Pro

### DISCUSSION THREAD

**Discussion by ThatStella7922** [ppy/osu#15721] [Ideas] [answered]
I've been looking into this now that building lazer for iOS works again, so I added the key to the `info.plist` and the game still is locked 60 FPS on iPhone 13 Pro. I've since learned that allowing 120FPS requires more work, like preferredFrameRateRange and possibly messing with a CADisplayLink. My experience with C# is too small to be able to do this properly so I'd like to hear ideas from others.

**Comment by peppy** [comment]
May be something @frenzibyte can take a look into. If it doesn't work out-of-the-box, it's likely that the vsync refresh rate is being artificially limited (or set incorrectly) somewhere in the stack. Sounds weird though, as this already works fine on iPads. I would've thought the plist change is enough.

**Comment by frenzibyte** [comment, marked as answer]
While I don't have a device with ProMotion display, I have tinkered around and found that osuTK already uses `CADisplayLink` as the timesource for updating the game, which is the reason why osu! always runs at 60 fps no matter what frame limiter mode was set.

The refresh rate options we currently have should probably not exceed 2x for iOS but that should be fixed separately.

I've updated osuTK's `CADisplayLink` usage to use the new frame rate range option and set it to Apple's suggested range for full-motion gaming (which I guess osu! qualifies to it, kinda?):
<img width="747" alt="CleanShot 2021-12-07 at 05 32 34@2x" src="https://user-images.githubusercontent.com/22781491/144955461-70dabebe-8ca8-43c5-8b4b-49a611411863.png">

(see https://developer.apple.com/documentation/quartzcore/optimizing_promotion_refresh_rates_for_iphone_13_pro_and_ipad_pro)

@ThatsNiceGuy can you checkout this branch (https://github.com/frenzibyte/osu-framework/tree/ios-high-frame-rate), run `SampleGame.iOS`, and see if it runs at 120 fps now? (for anyone wondering, the local osuTK changes are https://github.com/ppy/osuTK/compare/master..frenzibyte:ios-high-frame-rate, not 100% sure about it though, maybe should put the minimum to 60 rather than 80 for things to make more sense?)

**Reply by peppy** [reply]
For reference, osu! already runs at 120hz on ipads. But apparently the iOS (non iPadOS) spec is slightly different.

**Reply by frenzibyte** [reply]
Yeah, I'm suspecting this is because osuTK was using `PreferredFramesPerSecond = 0`, which is the same as `CAFrameRateRange.default` (0 min, 0 max, 0 preferred), and that range is considered "not higher frame rate" for iOS as in the table Apple referenced:

<img width="646" alt="CleanShot 2021-12-07 at 06 59 28@2x" src="https://user-images.githubusercontent.com/22781491/144963821-d0f4e656-27bc-4944-9b6e-d4b9b9e89d21.png">

It's also mentioned in the documentation I linked that these changes are not required for iPad, fwiw:

<img width="647" alt="CleanShot 2021-12-07 at 07 01 41@2x" src="https://user-images.githubusercontent.com/22781491/144964004-e7c13470-3ae5-47fe-9e56-b7e2ea76d8a8.png">

**Reply by ThatStella7922** [reply]
Just got started with this, apparently restoring packages for `osu.Framework.iOS` is trying to get version `2021.1207.0` of `ppy.osuTK.NS20` which doesn't exist.

<details>
 <summary>nuget log</summary>

```
Determining projects to restore...
/Users/thatsniceguy/Git/osu-framework-ios-high-frame-rate/osu.Framework.iOS/osu.Framework.iOS.csproj : error NU1102: Unable to find package ppy.osuTK.NS20 with version (>= 2021.1207.0) [/Users/thatsniceguy/Git/osu-framework-ios-high-frame-rate/osu-framework.sln]
/Users/thatsniceguy/Git/osu-framework-ios-high-frame-rate/osu.Framework.iOS/osu.Framework.iOS.csproj : error NU1102:   - Found 27 version(s) in nuget.org [ Nearest version: 1.0.178 ] [/Users/thatsniceguy/Git/osu-framework-ios-high-frame-rate/osu-framework.sln]
  Failed to restore /Users/thatsniceguy/Git/osu-framework-ios-high-frame-rate/osu.Framework.iOS/osu.Framework.iOS.csproj (in 589 ms).
  8 of 9 projects are up-to-date for restore.

  ```
</details>

Will check project files manually for any lingering references

**Reply by frenzibyte** [reply]
It's supposed to use local packages I've pushed to nuget for testing (`frenzi.osuTK.*`, see https://github.com/frenzibyte/osu-framework/commit/c28686f8474664a1b76c879995a2bdc29697494b).

**Reply by ThatStella7922** [reply]
did a bit more messing around and could not get packages to restore correctly, will try again tomorrow

**Reply by ThatStella7922** [reply]
trying again and I'm still having the issue where it's trying to pull `ppy.osuTK.NS20`

```
Feeds used:
    https://api.nuget.org/v3/index.json

Unable to find package ppy.osuTK.NS20 with version (>= 2021.1207.0)
  - Found 27 version(s) in nuget.org [ Nearest version: 1.0.178 ]
Restore failed.
```

tried pulling a fresh copy of the branch, so im not sure what's going on here

**Reply by frenzibyte** [reply]
That package should never be fetched from, it's `frenzi.osuTK.NS20` that should be... for reference, have you tried restoring from completely clean folder (i.e. remove all `bin` and `obj` folders in the framework solution)

**Reply by frenzibyte** [reply]
Will be resolved by:
 - https://github.com/ppy/osuTK/pull/71
 - https://github.com/ppy/osu/pull/15992

**Reply by KrisCris** [reply]
@frenzibyte Hey, just curious to ask - is it possible to force iOS to run at its full performance? I have an M1 iPad and the framerate on the level selection page hardly reaches even 70fps. But it is funny that, when I turned the screen recording on, it bumps to 120 immediately.

**Reply by frenzibyte** [reply]
Open a separate discussion for that.

**Comment by ThatStella7922** [comment]
solved, hits 120fps on iPhone 13 pro with frenzi's solution
